File: aoc-2024-rs/src/bin/day09.rs
use itertools::Itertools;

fn part1(disk_map: &str) {
    let mut extended_map: Vec<i64> = vec![];
    let mut file_block = true;
    let mut file_id = 0;
    for c in disk_map.chars() {
        if file_block {
            let file_size = c.to_digit(10).unwrap();
            extended_map.extend(vec![file_id; file_size as usize]);
            file_id += 1;
            file_block = false;
        } else {
            let block_size = c.to_digit(10).unwrap();
            extended_map.extend(vec![-1; block_size as usize]);
            file_block = true;
        }
    }
    let mut idx_file;
    loop {
        idx_file = (extended_map.len() - 1)
            - extended_map
                .iter()
                .rev()
                .find_position(|x| x >= &&0)
                .unwrap()
                .0;
        let idx_slot = extended_map.iter().find_position(|x| x < &&0).unwrap().0;
        // println!("{:?} {:?}", idx_file, idx_slot);
        if idx_file <= idx_slot {
            break;
        }
        extended_map.swap(idx_file, idx_slot);
    }
    // println!("{:?}", extended_map);
    let checksum = extended_map
        .iter()
        .take(idx_file + 1)
        .enumerate()
        .map(|(i, &x)| x as usize * i)
        .sum::<usize>();
    println!("{:?}", checksum);
}

fn part2(disk_map: &str) {
    let mut extended_map: Vec<i64> = vec![];
    let mut file_info: Vec<(i64, i64)> = vec![];
    let mut slot_info: Vec<(i64, i64)> = vec![];
    let mut file_block = true;
    let mut file_id = 0;
    for c in disk_map.chars() {
        if file_block {
            let file_size = c.to_digit(10).unwrap();
            let idx = extended_map.len();
            file_info.push((file_size as i64, idx as i64));
            extended_map.extend(vec![file_id; file_size as usize]);
            file_id += 1;
            file_block = false;
        } else {
            let block_size = c.to_digit(10).unwrap();
            let idx = extended_map.len();
            slot_info.push((block_size as i64, idx as i64));
            extended_map.extend(vec![-1; block_size as usize]);
            file_block = true;
        }
    }
    while file_id > 0 {
        file_id -= 1;
        let (file_size, file_idx) = file_info[file_id as usize];
        for i in 0..slot_info.len() {
            if slot_info[i as usize].1 > file_idx {
                break;
            }
            if slot_info[i as usize].0 >= file_size {
                for j in 0..file_size {
                    extended_map.swap(
                        file_idx as usize + j as usize,
                        slot_info[i as usize].1 as usize + j as usize,
                    );
                }
                slot_info[i as usize].0 -= file_size;
                slot_info[i as usize].1 += file_size;
                break;
            }
        }
    }
    // println!("{:?}", extended_map);
    let checksum = extended_map
        .iter()
        .enumerate()
        .map(|(i, &x)| if x > 0 { x as usize * i } else { 0 })
        .sum::<usize>();
    println!("{:?}", checksum);
}

fn main() {
    let input = include_str!("../../input/day09/in.txt");
    println!("{:?}", input);
    part1(input);
    part2(input);
}

File: aoc-2024-cpp/day09/day09.cpp
#include <algorithm>
#include <array>
#include <chrono>
#include <concepts>
#include <cstddef>
#include <cstdint>
#include <filesystem>
#include <fstream>
#include <iostream>
#include <iterator>
#include <map>
#include <optional>
#include <ranges>
#include <regex>
#include <string>
#include <string_view>
#include <vector>

#include <fmt/base.h>
#include <fmt/ranges.h>
#include <flux.hpp>

namespace sr = std::ranges;
namespace sv = std::views;
namespace fs = std::filesystem;

using fmt::print;
using fmt::println;

#define MEASURE(f)                                                             \
  {                                                                            \
    auto start = std::chrono::high_resolution_clock::now();                    \
    f;                                                                         \
    auto end = std::chrono::high_resolution_clock::now();                      \
    println("{} took {} µs", #f,                                               \
            std::chrono::duration_cast<std::chrono::microseconds>(end - start) \
                .count());                                                     \
  }

auto read_input(fs::path const& file_name) -> std::vector<std::string> {
  std::fstream infile(fs::path{BASE_PATH} / file_name);
  std::vector<std::string> lines;

  std::string line;
  while (std::getline(infile, line)) {
    lines.push_back(line);
  }
  infile.close();
  return lines;
}

void part1(std::string const& disk_map) {
  std::vector<std::int64_t> extended_map{};
  bool file_block = true;
  std::int64_t file_id = 0;
  for (auto c : disk_map) {
    if (file_block) {
      auto file_size = c - '0';
      for (int i = 0; i < file_size; ++i) {
        extended_map.push_back(file_id);
      }
      file_id++;
      file_block = false;
    } else {
      auto block_size = c - '0';
      for (int i = 0; i < block_size; ++i) {
        extended_map.push_back(-1);
      }
      file_block = true;
    }
  }

  auto ret = sr::partition(extended_map, [](auto x) { return x >= 0; });
  extended_map.erase(ret.begin(), ret.end());

  std::size_t checksum = 0;
  for (auto [i, x] : sv::enumerate(extended_map)) {
    checksum += i * x;
  }

  println("{}", checksum);
}

void part2(std::string const& disk_map) {
  std::vector<std::int64_t> extended_map{};
  std::vector<std::pair<std::size_t, std::size_t>> file_info{};
  std::vector<std::pair<std::size_t, std::size_t>> slot_info{};
  bool file_block = true;
  std::int64_t file_id = -1;
  for (auto c : disk_map) {
    if (file_block) {
      file_id++;
      auto file_size = c - '0';
      auto idx = extended_map.size();
      for (int i = 0; i < file_size; ++i) {
        extended_map.push_back(file_id);
      }
      file_info.emplace_back(file_size, idx);
      file_block = false;
    } else {
      auto block_size = c - '0';
      auto idx = extended_map.size();
      for (int i = 0; i < block_size; ++i) {
        extended_map.push_back(-1);
      }
      slot_info.emplace_back(block_size, idx);
      file_block = true;
    }
  }
  while (file_id >= 0) {
    auto [file_size, file_idx] = file_info.at(file_id);
    if (auto it = sr::find_if(slot_info,
                              [&](auto s) {
                                return (s.first >= file_size) &&
                                       (s.second < file_idx);
                              });
        it != slot_info.end()) {
      auto [slot_size, slot_idx] = *it;
      for (int i = 0; i < file_size; ++i) {
        sr::iter_swap(extended_map.begin() + file_idx + i,
                      extended_map.begin() + slot_idx + i);
      }
      it->first -= file_size;
      it->second += file_size;
    }
    file_id--;
  }
  std::size_t checksum = 0;
  for (auto [i, x] : sv::enumerate(extended_map)) {
    if (x >= 0) {
      checksum += i * x;
    }
  }
  println("{}", checksum);
}

int main() {
  using namespace std::literals;
  auto lines = read_input("day09/in.txt");
  auto disk_map = lines[0];
  println("{}", disk_map);
  MEASURE(part1(disk_map))
  MEASURE(part2(disk_map))
}
